### PR TITLE
[KOGITO-5446] Introducing Kogito api and client modules

### PR DESCRIPTION
**Pull request by ricardozanini**
Signed-off-by: Ricardo Zanini <zanini@redhat.com>

See: https://issues.redhat.com/browse/KOGITO-5446

In this PR we add two new modules to the Kogito Operator project: `api` and `client`. This way, third parties can leverage our custom resources and interface with them via the Kubernetes API. Of course, the Kogito Operator installation is a requirement for such use cases. 

```go
kogitoClient := clientset.NewForConfig(kubernetesRestConfig)
kogitoSvc, err :=
		kogitoClient.AppV1beta1().KogitoRuntimes("kogito").Get(context.TODO(), "myservice", metav1.GetOptions{})

// do whatever you want with the KogitoRuntime resource 
```

TODO:

- [x] Documentation
- [x] Release Notes
- [x] Add "enable/disable" API module script
- [x] Fix CI
- [x] Fix version bump to consider client-gen
- [x] Add client-gen to other generators (add to makefile as well)
- [x] Unit tests with client-gen

Please be aware that the entire `client` module is generated, excepted by the `client_test.go` file. Don't bother reviewing it. :sweat: 

Many thanks for submitting your Pull Request :heart:! 

Please make sure your PR meets the following requirements:

- [x] You have read the [contributors' guide](https://github.com/kiegroup/kogito-operator/blob/master/README.md#contributing-to-the-kogito-operator)
- [x] Pull Request title is properly formatted: `[KOGITO-XYZ] Subject`
- [x] Pull Request contains a link to the JIRA issue
- [x] Pull Request contains a description of the issue
- [x] Pull Request does not include fixes for issues other than the main ticket
- [x] Your feature/bug fix has a unit test that verifies it
- [ ] You've tested the new feature/bug fix in an actual OpenShift cluster
- [x] You've added a [RELEASE_NOTES.md](RELEASE_NOTES.md) entry regarding this change

### CODE REVIEW

**Comment by ricardozanini** [comment]
OLM error does not seem to be related to the changes in this PR:

```
2021-06-30 21:12:36,490 docker_builder.py:141        ERROR You can look inside the failed image by running 'docker run --rm -ti 6005d6a313d7 bash'
2021-06-30 21:12:36,490 cli.py:342        ERROR ('Image build failed, see logs above.', CekitError("Image build failed: 'failed to export image: failed to create image: failed to get layer sha256:92274b9fe76929f88a0820cc9cae5609163667adb7aecfcc749a73da3b5fd8e0: layer does not exist'"))
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/dist-packages/cekit/builders/docker_builder.py", line 89, in _build_with_docker
    raise CekitError("Image build failed: '{}'".format(error_message))
cekit.errors.CekitError: Image build failed: 'failed to export image: failed to create image: failed to get layer sha256:92274b9fe76929f88a0820cc9cae5609163667adb7aecfcc749a73da3b5fd8e0: layer does not exist'
```
Tomorrow I'll take a look, but it looks like an error in the CI env. Runs perfectly locally.

**Comment by ricardozanini** [comment]
Problem with the image build fixed.

**Review by sutaakar — APPROVED** [review]
Looks good to me, though I am not much sure which parts are generated and which are added manually.

**Comment by ricardozanini** [comment]
> Looks good to me, though I am not much sure which parts are generated and which are added manually.

the `client` package is entirely generated. Plus you can see a comment in every file:

```
// Code generated by client-gen. DO NOT EDIT.
```

File: client/clientset/versioned/scheme/register.go [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+//
+//   kclientset, _ := kubernetes.NewForConfig(c)
+//   _ = aggregatorclientsetscheme.AddToScheme(clientsetscheme.Scheme)
+//
+// After this, RawExtensions in Kubernetes types will serialize kube-aggregator types
+// correctly.
+var AddToScheme = localSchemeBuilder.AddToScheme
+
+func init() {
+	v1.AddToGroupVersion(Scheme, schema.GroupVersion{Version: "v1"})
```
Comment (spolti) [inline client/clientset/versioned/scheme/register.go:51]: shouldn't be alpha?

**Review by spolti — APPROVED** [review]
Very nice addition.
small questions, not mandatory though :)

File: README.md [hunk 2]
```diff
@@ -59,13 +56,34 @@ $ ./hack/install.sh
...
+Replace `{VERSION}` with the [latest release](https://github.com/kiegroup/kogito-operator/releases).
+
+The `client` module will provide your project with the Kubernetes client 
+for Kogito custom resources such as `KogitoRuntime` and `KogitoBuild`.
+
+Avoid importing the whole project unless extremely necessary. The Kogito Operator
+has many dependencies since it can connect to different managed resources
+from third-party vendors such as Strimzi and Infinispan.
+
+# Contributing to the Kogito Operator
```
[the remark this answers is not in the document]
Comment (ricardozanini) [reply README.md:82]: This is a work @Kevin-Mok is doing. Once is done, we can update this section.

File: client/clientset/versioned/scheme/register.go (hunk 1 above)
Comment (ricardozanini) [reply client/clientset/versioned/scheme/register.go:51]: Absolutely, this is the register of the core k8s objects.

File: hack/kogito-module-api.sh [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+  echo "This script handles Kogito API module manipulation with Operator SDK"
+  echo "Must run from the project root"
+  echo "USAGE: ./kogito-api.sh [--disable|--enable]"
+  echo "--disable disables the Kogito API module"
+  echo "--enable  enables the Kogito API module"
+}
+
+disable() {
+  sed -i '/github.com\/kiegroup\/kogito-operator\/api v/d' ./go.mod
+	mv ./api/go.mod api/go.mod.bkp 2>/dev/null; true
```
[the remark this answers is not in the document]
Comment (ricardozanini) [reply hack/kogito-module-api.sh:29]: Weird, in my editor, it is correct. I'll run the formatter again.

**Comment by vaibhavjainwiz** [comment]
@ricardozanini @sutaakar @spolti 
Now we have three go module(api/client/operator) and all of them needed to be release seperately.
How we are going to maintain dependency between those?
**Does we always release all three go module togeather?**
like after this sprint we have to release 1.8.0 version, so does we release them all togeather.

- kogito-operator/api(v1.8) 
- kogito-operator/client(v1.8) 
- kogito-operator(v1.8)

If we release them all togeather, does it mean if user upgrade its `kogito-operator` version from v1.7 to v1.8, then he have to make changes in it application code to use kogito-operator/api(v1.8) and kogito-operator/client(v1.8) to maintain compatiblity as we recommended.

**Or we will release them seperately?**
- kogito-operator(v1.8)
- kogito-operator/api(v1.0)
- kogito-operator/client(v1.0)

If we release them seperately then we could only release  kogito-operator/api & kogito-operator/client, if there are some changes. So if user upgrade its `kogito-operator` from v1.7 to v1.8 and there is no changes in api's then he don't need to make any change in its code. He could keep using same api/client go module in its codebase. 
If we are planning to release them seperatly then we have to provide compatiblity metric in kogito-doc.

**Comment by sutaakar** [comment]
I think it may be better to release all modules together. That way we don't have to worry about compatibility between components.

**Comment by ricardozanini** [comment]
@vaibhavjainwiz, we don't need to change anything in the release process. These modules are part of the kogito-operator main project and won't be released separately by any means. I understand that you approved this PR? I intend to merge it today.

**Review by vaibhavjainwiz — APPROVED** [review]

**Pull request merged**